packages/loader/container-loader/src/retriableDocumentStorageService.ts

```diff
@@ -0,0 +1,95 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { CreateContainerError } from "@fluidframework/container-utils";
+import { IDocumentStorageService } from "@fluidframework/driver-definitions";
+import { canRetryOnError, DocumentStorageServiceProxy } from "@fluidframework/driver-utils";
+import { ISnapshotTree, IVersion } from "@fluidframework/protocol-definitions";
+import { Container, RetryFor } from "./container";
+import { getRetryDelayFromError } from "./deltaManager";
+
+export class RetriableDocumentStorageService extends DocumentStorageServiceProxy {
+    private static callsWaiting: number = 0;
+    private static futureTimeTillWait: number = 0;
+    private disposed = false;
+    constructor(
+        internalStorageService: IDocumentStorageService,
+        private readonly container: Pick<Container, "emitDelayInfo" | "cancelDelayInfo">,
+    ) {
+        super(internalStorageService);
+    }
+
+    public dispose() {
+        this.disposed = true;
+    }
+
+    public async getSnapshotTree(version?: IVersion): Promise<ISnapshotTree | null> {
+        return this.readWithRetry(async () => this.internalStorageService.getSnapshotTree(version));
+    }
+
+    public async read(blobId: string): Promise<string> {
+        return this.readWithRetry(async () => this.internalStorageService.read(blobId));
+    }
+
+    public async readBlob(id: string): Promise<ArrayBufferLike> {
+        return this.readWithRetry(async () => this.internalStorageService.readBlob(id));
+    }
+
+    public async readString(id: string): Promise<string> {
+        return this.readWithRetry(async () => this.internalStorageService.readString(id));
+    }
+
+    private async delay(timeMs: number): Promise<void> {
+        return new Promise((resolve) => setTimeout(() => resolve(), timeMs));
+    }
+
+    private async readWithRetry<T>(api: () => Promise<T>): Promise<T> {
+        let result: T | undefined;
+        let success: boolean | undefined;
+        let retryAfter = 0;
+        do {
+            try {
+                result = await api();
+                if (success === false) {
+                    RetriableDocumentStorageService.callsWaiting -= 1;
+                    if (RetriableDocumentStorageService.callsWaiting === 0) {
+                        RetriableDocumentStorageService.futureTimeTillWait = 0;
+                        this.container.cancelDelayInfo(RetryFor.Storage);
+                    }
+                }
+                success = true;
+            } catch (err) {
+                // If it is not retriable, then just throw the error.
+                const canRetry = canRetryOnError(err);
+                if (!canRetry || this.disposed) {
+                    throw err;
+                }
+                if (success === undefined) {
+                    // We are going to retry this call.
+                    RetriableDocumentStorageService.callsWaiting += 1;
```

Comment:
Can we get rid of callsWaiting  somehow?
Here is a suggestion - it does what RetryFor is doing at fixed (3) scope - introducing N number of independent waiters.
Can we generalize RetryFor mechanism? I.e. remove 3 fixed entries, and instead have a variable slot system, where users can acquire slot (index), keep updating slot, and then release slot.
Or maybe slot can be some unique key (string) that is generated by callers where callers somehow guarantee uniqueness of it (like guid). Then whole system becomes much simpler and more unified.

Reply:
Removed calls waiting. implemented a unique id system(guid).